controller/volume_controller.go

```diff
@@ -2863,29 +2863,43 @@ func (vc *VolumeController) checkAndFinishVolumeRestore(v *longhorn.Volume, e *l
 		!v.Spec.Standby) {
 		return nil
 	}
-	allScheduledReplicasIncluded := true
+
+	allScheduledReplicasIncluded, err := vc.checkAllScheduledReplicasIncluded(v, e, rs)
+	if err != nil {
+		return err
+	}
+
+	if !isPurging && (v.Status.Robustness == longhorn.VolumeRobustnessHealthy || v.Status.Robustness == longhorn.VolumeRobustnessDegraded) && allScheduledReplicasIncluded {
+		log.Info("Restore/DR volume finished")
+		v.Status.IsStandby = false
+		v.Status.RestoreRequired = false
+	}
+
+	return nil
+}
+
+func (vc *VolumeController) checkAllScheduledReplicasIncluded(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) (bool, error) {
+	healthReplicaCount := 0
+	hasReplicaNotIncluded := false
+
 	for _, r := range rs {
 		// skip unscheduled replicas
 		if r.Spec.NodeID == "" {
 			continue
 		}
 		if isDownOrDeleted, err := vc.ds.IsNodeDownOrDeleted(r.Spec.NodeID); err != nil {
-			return err
+			return false, err
 		} else if isDownOrDeleted {
 			continue
 		}
-		if mode := e.Status.ReplicaModeMap[r.Name]; mode != longhorn.ReplicaModeRW {
-			allScheduledReplicasIncluded = false
-			break
+		if mode := e.Status.ReplicaModeMap[r.Name]; mode == longhorn.ReplicaModeRW {
+			healthReplicaCount++
+		} else {
+			hasReplicaNotIncluded = true
 		}
 	}
-	if !isPurging && (v.Status.Robustness == longhorn.VolumeRobustnessHealthy || v.Status.Robustness == longhorn.VolumeRobustnessDegraded) && allScheduledReplicasIncluded {
-		log.Info("Restore/DR volume finished")
-		v.Status.IsStandby = false
-		v.Status.RestoreRequired = false
-	}
 
-	return nil
+	return healthReplicaCount != 0 && (healthReplicaCount >= v.Spec.NumberOfReplicas || !hasReplicaNotIncluded), nil
 }
```

Comment:
Previous implementation will directly return `allScheduledReplicasIncluded is false` if any replica failed even there are `v.Spec.NumberOfReplicas or more` healthy replicas.

`healthReplicaCount == v.Spec.NumberOfReplicas` will not handle if there are the `NumberOfReplicas + n` healthy replicas
I think `! hasReplicaNotIncluded` is implicitly equal to `(areAllReplicasIncluded && healthReplicaCount < v.Spec.NumberOfReplicas)` because we have checked the `healthReplicaCount >= v.Spec.NumberOfReplicas`

Yes, I had tested this PR and passed the [reproducing steps](https://github.com/longhorn/longhorn/issues/3069#issue-1007296119)

Reply:
Sounds good.